pkg/credentialprovider/gcp/metadata.go

```diff
@@ -142,7 +145,17 @@ func onGCEVM() bool {
 
 // Enabled implements DockerConfigProvider for all of the Google implementations.
 func (g *MetadataProvider) Enabled() bool {
-	return onGCEVM()
+	if credentialprovider.AreLegacyCloudCredentialProvidersDisabled() {
```

Comment:
Not sure I agree with this ordering. I agree about the relative costs but if we need to enter either switch the system is miconfigured. At the point the cost seems like a non issue. For the good case you need to pass both switched anyway. In this case we may have a misleading error message suggesting you are on GCP. I think it would be better to put back the original order and add a warning you are attempting to use a GCP credential provider on a non GCP VM to the !onGCEVM case.

Reply:
Avoiding the misleading error message was my original thinking, happy to switch it back.

I'm not sure if the other warning you propose would add much value though: IIUC these credential providers are on by default (thus the need for this PR) so it's not like the user misconfigured something if we're getting this far on a non-GCP VM. The warning would just get logged all the time and wouldn't be actionable.

Reply:
I thought we would only have this on if it was configured. If thats not true, then I think its even more important to switch back. (As you say to avoid the misleading error message)

Reply:
Switched this back around.

Reply:
I'm not following the logic here to switch the order back to run `onGCEVM` first. If the feature gate is enabled (i.e. disabled), then there's no reason to ever call onGCEVM, because the plugin is explicitly disabled. In this case, the warning message makes sense too, because the plugin is disabled anyways. When it is disabled, why bother running  `onGCEVM` which is a relatively expensive operation that reads a file from disk?

Reply:
With the in-tree cloud providers, you have to pass `--cloud-provider=foo` to enable them so it makes sense to print a warning if you pass that flag while the feature gate that disables them is turned on, since that's obviously a misconfiguration of either the flag or the feature gate.

With the in-tree credential providers OTOH, there's no flag to enable them (as far as I'm aware) -- we just rely on the `onGCEVM` check alone. So if the feature gate is turned on but we're not on a GCP VM, there's no reason to print a warning about the GCP provider being disabled... Nothing is misconfigured in that case, the warning is just confusing and unactionable.

Reply:
Gotcha, thanks for clarifying.